### ISSUE TITLE
Week 1 — Manual Line Learner: watch numbers learn

### ISSUE THREAD

**Issue by majorgilles** [majorgilles/intuitive_understanding_machine_learning#2]
This issue is part of an 8-week self-study course for understanding how machines work and learn by building and inspecting small local ML projects.

**Type:** HITL  
**Labels:** `HITL`, `course`, `learning`, `beginner`, `weekly-project`  
**Time estimate:** 4–6 hours  
**Blocked by:** #1

### Project name

**Manual Line Learner**

### Why this exists

Frameworks can hide the learning process. This project makes learning visible by using plain Python/NumPy first.

### What to build

Build a tiny model that learns the rule:

```text
y = 2x + 1
```

The model should start with bad guesses for two knobs:

- `weight`
- `bias`

Then it should repeatedly improve those knobs and show the loss going down.

### Dataset

Generate the data inside the script.

Use this exact clean dataset first:

```python
x = [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5]
y = 2 * x + 1
```

No download needed.

### Exact files to create

- `weeks/week-01-learning-loop/manual_line_learner.py`
- `weeks/week-01-learning-loop/notes.md`
- `artifacts/week-01-learning-loop/loss_curve.png`
- `artifacts/week-01-learning-loop/before_after_predictions.csv`

### Online resources

- ML4A: How neural networks are trained: https://ml4a.github.io/ml4a/how_neural_networks_are_trained/
- TensorFlow Neural Network Playground: https://playground.tensorflow.org/
- Google Machine Learning Crash Course: https://developers.google.com/machine-learning/crash-course

### Step-by-step work

- [x] Create the dataset in code.
- [x] Start with intentionally wrong values, for example `weight = 0.0` and `bias = 0.0`.
- [x] Write a `predict(x, weight, bias)` function.
- [x] Write a simple loss function: average squared wrongness.
- [x] Print the first predictions before learning.
- [x] Run a training loop for about 100–500 steps.
- [x] At each step, adjust `weight` and `bias` a little.
- [x] Save loss values.
- [x] Plot the loss curve to `loss_curve.png`.
- [x] Save before/after predictions to `before_after_predictions.csv`.
- [x] Write `notes.md` explaining what changed and why loss went down.

### Important constraint

Do **not** use PyTorch autograd in this issue. The point is to see the adjustment process without framework magic.

### Modification challenge

Change the hidden rule to:

```text
y = -3x + 2
```

Before running, write a prediction in `notes.md`:

- What should the final `weight` become?
- What should the final `bias` become?

Then run it and compare.

### Debugging checklist

- [ ] If loss goes up, the adjustment direction may be backwards.
- [ ] If loss changes too slowly, the step size may be too small.
- [ ] If loss explodes, the step size may be too large.
- [ ] If final predictions are bad, print the first five predictions every 50 steps.

### Acceptance criteria

- [ ] `manual_line_learner.py` runs with `uv run python weeks/week-01-learning-loop/manual_line_learner.py`.
- [ ] The script prints starting and ending `weight` and `bias`.
- [ ] Final `weight` and `bias` are close to the hidden rule.
- [ ] `loss_curve.png` exists and shows loss going down.
- [ ] `before_after_predictions.csv` exists.
- [ ] `notes.md` explains the learning loop in plain words.

### Expertise gained

After this issue, I can explain machine learning as repeated correction of adjustable knobs.

---
